core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala

```diff
@@ -1408,4 +1404,239 @@ class ReplicaManagerTest {
+
+    val controllerEpoch = 0
+    var leaderEpoch = 0
+    val leaderEpochIncrement = 1
+    val correlationId = 0
+    val controllerId = 0
+
+    val partition0Replicas0 = Seq[Integer](0).asJava
+    val partition0Replicas01 = Seq[Integer](0, 1).asJava
+    val partition0Replicas1 = Seq[Integer](1).asJava
+
+    // make broker 0 the leader of partition 0
+    val leaderAndIsrRequest1 = new LeaderAndIsrRequest.Builder(ApiKeys.LEADER_AND_ISR.latestVersion,
+      controllerId, 0, brokerEpoch,
+      Seq(new LeaderAndIsrPartitionState()
+        .setTopicName(tp0.topic())
+        .setPartitionIndex(tp0.partition())
+        .setControllerEpoch(controllerEpoch)
+        .setLeader(0)
+        .setLeaderEpoch(leaderEpoch)
+        .setIsr(partition0Replicas0)
+        .setZkVersion(0)
+        .setReplicas(partition0Replicas0)
+        .setIsNew(true)).asJava,
+      Set(new Node(0, "host0", 0), new Node(1, "host1", 1)).asJava).build()
+
+    rm.becomeLeaderOrFollower(correlationId, leaderAndIsrRequest1, (_, _) => ())
+
+    // Append a couple of messages.
+    for (i <- 1 to 2) {
+      appendRecords(rm, tp0, TestUtils.singletonRecords(s"message $i".getBytes)).onFire { response =>
+        assertEquals(Errors.NONE, response.error)
+      }
+    }
+
+    // put rm0 and rm1 into reassignment state
+    leaderEpoch += leaderEpochIncrement
+    val leaderAndIsrRequest2 = new LeaderAndIsrRequest.Builder(ApiKeys.LEADER_AND_ISR.latestVersion, controllerId,
+      controllerEpoch, brokerEpoch,
+      Seq(new LeaderAndIsrPartitionState()
+        .setTopicName(tp0.topic())
+        .setPartitionIndex(tp0.partition())
+        .setControllerEpoch(controllerEpoch)
+        .setLeader(0)
+        .setLeaderEpoch(leaderEpoch)
+        .setIsr(partition0Replicas0)
+        .setZkVersion(0)
+        .setReplicas(partition0Replicas01)
+        .setAddingReplicas(partition0Replicas1)
+        .setIsNew(true)).asJava,
+      Set(new Node(0, "host0", 0), new Node(1, "host1", 1)).asJava).build()
+
+    rm.becomeLeaderOrFollower(correlationId, leaderAndIsrRequest2, (_, _) => ())
+
+    // after the fetch the lag should be increased to 2 as 2 messages have been produced
+    val fetchResult = fetchAsFollower(rm, tp0, new PartitionData(0, 0, Int.MaxValue, Optional.empty()))
+    var timeout = 1000L
+    while (!fetchResult.isFired && timeout > 0) {
+      wait(100L)
+      timeout -= 100
+    }
+    assertEquals(2, rm.fetchersMaxLag())
+
+    // we have the lag now and let's simulate that the reassignment ended where tp0 is deleted
+    // so we can remove the lag stats
+    rm.maybeUpdateMetadataCache(0, new UpdateMetadataRequest.Builder(0,0,0,0,
+      Collections.emptyList(), Collections.emptyList()).build(ApiKeys.UPDATE_METADATA.latestVersion()))
+
+    assertEquals(0, rm.fetchersMaxLag())
+  }
+
+  private def prepareDifferentReplicaManagers(): (ReplicaManager, ReplicaManager) = {
```

Comment:
Could we reduce the duplication of this method with `prepareDifferentReplicaManagersWithMockedBrokerTopicStats`?

`prepareDifferentReplicaManagers` can accept two `BrokerTopicStats` which default to the real thing and the `prepareDifferentReplicaManagersWithMockedBrokerTopicStats` could pass in mocks